### PR TITLE
KT-8263: Conditional operators are not parsed correctly

### PR DESCRIPTION
**Pull request by kralliv**
The parsing of type argument lists now differentiates depending on the grammatical environments they appear in. There are two environments: strict and mixed. In strict environments there cannot be any grammar interactions with comparison expressions, whereas in mixed environments such interactions need to be accounted for. Strict environments can be found pretty much everywhere except in blocks/lambdas. Examples for strict environments are type references in properties or value parameter list of functions. Inside of blocks/lambdas the mixed environment allows for interactions between call expressions such as `mutableListOf<Here>()` and comparison such as `some(x < y, 0 > (y + z))`.

It is crucial, that the parser always produces a type argument list, if there is a grammatically correct call expression, even in cases where it could also be parsed as a grammatically correct comparison. The call expression takes precedence over the comparison, as a call expression cannot be written in any other way, unlike comparisons, which can easily be expressed differently. On the other hand, a type argument list must not be produced, if there is a grammatically incorrect function call, but a grammatically correct comparison. In that case, a comparison has to be produced. If a section is neither a grammatically correct call expression nor comparison, then the parser is free to chose whatever option it assumes fits best. The first two cases are required for correctness. The third case is important for helpful error highlighting.

```kotlin
some(x < y, 0 > (y + z))   // comparison
some(x < y, x > (y + z))   // function call
some(x < y,  (y + z) < x)  // rewritten comparison
some(x < y, () > (y + z))  // neither
```

### Algorithm

Whether the parser produces a type argument list depends on the following conditions.

`KotlinParsing.tryParseTypeArgumentList(boolean)` parses a section and assigns it one of three kinds:
 - `TYPE_ARGUMENT_LIST` if the environment is strict, the type argument list is empty `<>` or if the type argument list did not contain any syntactical errors. This case implies, that the section is most likely a type argument list.
 - `FAULTY_TYPE_ARGUMENT_LIST` if none of the above conditions apply and the type argument list has a GT at the end, meaning that it is at least complete `<...>`. This case implies, that the section could be type argument list, that simply contains errors. But it could also be some sort of comparison. Further checks have to be done in this case.
 - `NONE` if none of the above conditions apply. As the name implies, the section cannot be considered a type argument list.

`KotlinExpressionParsing.parseCallSuffixTypeArgumentList()` decides based upon this determined kind and type argument list indicators (see below) whether a type argument list should be produced or not.
 - If the kind is `NONE`, **no type argument list will be produced**.
 - If the kind is `FAULTY_TYPE_ARGUMENT_LIST` the parser has to check whether the section could also be correct comparison. In order to do this, it rolls back the type argument list and tries to parse the section as the right-hand side of a comparison. If there are no syntax errors, **no type argument list will be produced**. Otherwise the right-hand side of the comparison will be rolled back, parsed as a type argument list once more and **a type argument list will be produced**. This whole check is only done, if the section is followed by a strong type argument list indicator.
 - If the kind is `TYPE_ARGUMENT_LIST` and if the sections is followed by any type argument list indicator, **a type argument list will be produced**.

### Type Argument List Indicators

In order to decide whether a type argument list should be produced, the tokens after the section have to be inspected. There are a number of indicators, that either clearly state the necessity of the production of a type argument list or at least hint towards the intention of such existence. The former indicators are referred to as strong, whereas the latter are considered weak.

#### Strong Type Argument List Indicators

In order of precedence:

1. The start of a value argument list or an annotated lambda indicates an intended type argument list.
`<...>(` `<...>{` `<...>@identifier{` `<...>identifier@{`
2. Postfix expressions are usually not applied to non-null boolean comparisons.
`<...>::` `<...>?` `<...>.` `<...>?.` `<...>!!` `<...>++` `<...>--`

#### Weak Type Argument List Indicators

1. Line breaks are only available outside of value argument list. This is an incomplete line of code.
`<...>↵`
2. Other limiting tokens are also a likely indicator of an indented function call.
`<...>)` `<...>]` `<...>}` `<...>,` `<...>:` `<...>=`
3. Operators are unlikely to be applied to non-null booleans.
`<...>==` `<...>!=` `<...>===` `<...>!==` `<...> as` `<...> is` `<...> in`
4. Hard keywords hint towards some formatting issue where line breaks are missing
`<...>fun` `<...>var` `<...>var` `<...>object`

### Coverage

The current approach covers all cases when it comes to correctness. If the code compiles without errors, then the parser will decide correctly. It also covers a significant part of the error spectrum, aiding the developer in understanding their mistakes.

### Other

KT-35811 as covered in #4771 is re-included in this PR, because of the previous rollback in relation with KT-8263.

### CODE REVIEW

File: compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinParsing.java [hunk 1]
```diff
@@ -885,12 +895,10 @@ assert _at(IDENTIFIER) ||
 
         PsiBuilder.Marker reference = mark();
         PsiBuilder.Marker typeReference = mark();
-        parseUserType(/* allowSimpleIntersectionTypes */ false);
+        parseUserType(/* allowSimpleIntersectionTypes */ false,  /* strictTypeParameters */ true);
         typeReference.done(TYPE_REFERENCE);
         reference.done(CONSTRUCTOR_CALLEE);
 
-        parseTypeArgumentList();
```
Comment (kralliv) [inline compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinParsing.java:892]: I also discovered this little "feature" which allows writing thing like:
```kotlin
@Target(AnnotationTarget.FUNCTION)
@Retention(AnnotationRetention.SOURCE)
annotation class Ann<T>

@Ann<String><Here.You.Can, Write.Anything, ()->Unit, As.Long.As.It.Looks.Like.A.Type>
fun f() {}
```

**Review by dzharkov — CHANGES_REQUESTED** [review]
First of all, I have to say that I'm really impressed by the scale and the quality of work you've done here.
But as you've noticed from the destiny of KT-35811, parsing is a really fragile and complicated subsystem, so we have to pay a lot of attention when considering changing it even a bit.
Thus I left some clarification questions there.

**Comment by kralliv** [comment]
I had to update the approach to cover to the error cases. This far, any error in the type argument list would result in it being dropped. Therefor error highlighting and autocompletion would have been very spotty while filling out the type argument list of call expression. Writing calls like `mutableListOf<()->>()`, where the return type of the function type hasn't been typed yet would not receive proper highlighting. The behaviour should now be significantly smoother with this adjusted approach.

The changes primarily focus on `KotlinParsing.tryParseTypeArgumentList(boolean)` and `KotlinExpressionParsing.parseCallSuffixTypeArgumentList()`. I updated the description in the PR accordingly.

File: compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java [hunk 2]
```diff
@@ -59,17 +59,15 @@ private static ImmutableMap<String, KtToken> tokenSetToMap(TokenSet tokens) {
...
     }
 
-    private static final TokenSet TYPE_ARGUMENT_LIST_INTENTION_INDICATORS = TokenSet.orSet(
-            Precedence.POSTFIX.getOperations(),
-            TokenSet.create(
-                    RBRACKET, RBRACE, RPAR, EQ, COLON, COLONCOLON, SEMICOLON, COMMA,
+    private static final TokenSet TYPE_ARGUMENT_LIST_INTENTION_INDICATORS = TokenSet.create(
+            RBRACKET, RBRACE, RPAR, EQ, COLON, SEMICOLON, COMMA,
 
-                    EQEQEQ, EXCLEQEQEQ, EQEQ, EXCLEQ, SAFE_ACCESS, ELVIS, AS_KEYWORD, IS_KEYWORD, NOT_IS, NOT_IN, RANGE, RANGE_UNTIL,
```
[the remark this answers is not in the document]
Comment (kralliv) [reply compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java]: In the updated approach it should be a lot clearer. `SAFE_ACCESS` was removed from that list, because it is already contained in `Precedence.POSTFIX.getOperations()`. It is now part of the `STRONG_TYPE_ARGUMENT_LIST_INDICATORS`.

File: compiler/tests/org/jetbrains/kotlin/jvm/compiler/CompileEnvironmentTest.java [hunk 3]
```diff
@@ -32,6 +32,20 @@
 
 public class CompileEnvironmentTest extends TestCase {
 
+    public static void main(String[] args) {
```
[the remark this answers is not in the document]
Comment (kralliv) [reply compiler/tests/org/jetbrains/kotlin/jvm/compiler/CompileEnvironmentTest.java]: Oh, that was a completely unrelated test of mine. I removed it.

File: compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java [hunk 4]
```diff
@@ -489,30 +485,82 @@ else if (atSet(Precedence.POSTFIX.getOperations())) {
...
+                typeArgumentList.rollbackTo();
+
+                return false;
+            }
+
+            // The definite cases were the call-suffix takes precedence
+
+            // <...> (
+            if (isAtValueArgumentList()) {
                 typeArgumentList.done(TYPE_ARGUMENT_LIST);
```
[the remark this answers is not in the document]
Comment (kralliv) [reply compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java]: It depends heavily on what the section is followed by whether or not we can consider it as a call expression.
```kotlin
some(x < y, z > y)   // comparison
some(x < y, z > (y)) // call expression
some(x < y, z > { }) // call expression
```

File: compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java [hunk 5]
```diff
@@ -489,30 +485,82 @@ else if (atSet(Precedence.POSTFIX.getOperations())) {
...
+                typeArgumentList.done(TYPE_ARGUMENT_LIST);
+
+                parseCallWithClosure();
+                return true;
             }
-        }
-        else {
+
+            // <...>::
+            if (at(COLONCOLON)) {
```
[the remark this answers is not in the document]
Comment (kralliv) [reply compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java]: I'm going to a lot of test cases for all sorts of corner cases. I just didn't add them yet, because I wanted to work out the details of the approach first, without having to adjust the ranges within the tests over and over.

File: compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinParsing.java [hunk 6]
```diff
@@ -2423,43 +2446,58 @@ boolean tryParseTypeArgumentList(TokenSet extraRecoverySet) {
...
     }
 
     /*
      * functionType
      *   : (type ".")? "(" parameter{","}? ")" "->" type?
      *   ;
      */
-    private void parseFunctionType(PsiBuilder.Marker functionType) {
-        parseFunctionTypeContents(functionType).done(FUNCTION_TYPE);
+    private void parseFunctionType(PsiBuilder.Marker functionType, TokenSet recoverySet) {
```
[the remark this answers is not in the document]
Comment (kralliv) [reply compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinParsing.java:2554]: It was somewhat unimportant in the original approach, but has now become more important in the updated approach. What it does, is returning to `tryParseTypeArgumentList` sooner without having consumed as many tokens as it would have. If the token is contained in the `TYPE_ARGUMENT_LIST_RECOVERY_SET`, it will abort parsing the type argument list and return `TypeArgumentListKind.NONE`.

File: compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java [hunk 7]
```diff
@@ -489,30 +485,82 @@ else if (atSet(Precedence.POSTFIX.getOperations())) {
...
-        else if (at(LPAR)) {
-            parseValueArgumentList();
-            parseCallWithClosure();
-        }
-        else if (at(LT)) {
+        if (at(LT)) {
             PsiBuilder.Marker typeArgumentList = mark();
-            if (myKotlinParsing.tryParseTypeArgumentList(TYPE_ARGUMENT_LIST_STOPPERS)) {
+
+            if (!myKotlinParsing.tryParseTypeArgumentList(/* strict */false)) {
```
[the remark this answers is not in the document]
Comment (kralliv) [reply compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java]: I improved the documentation in the code. If is is too excessive, we can dial it back.

File: compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinParsing.java [hunk 8]
```diff
@@ -2423,43 +2446,58 @@ boolean tryParseTypeArgumentList(TokenSet extraRecoverySet) {
...
-            if (at(GT)) {
-                break;
+
+            if (at(COMMA)) {
+                advance(); // COMMA
+            }
+            else if (!at(GT) && !at(GTEQ)) {
+                // if it is not strict, then we want to fail fast as any error
+                // will roll back this type argument list anyway
+                if (!strict || atSet(TYPE_ARGUMENT_LIST_RECOVERY_SET) || eof()) break;
```
[the remark this answers is not in the document]
Comment (kralliv) [reply compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinParsing.java]: I added documentation to `tryParseTypeArgumentList` along side with the explanation of the returned values and their conditions.

I completely agree, that we should not introduce unnecessary complexity into an already extensive change. It originally sought to improved error recovery in strict environments only, which was unrelated to the problem itself, but it has now become a critical part in the updated approach. We now always try to reach the GT, independent of the environment. If we cannot reach the GT without running into a token from the recovery set, then we consider the parsed section as not a type argument list, similar to how it was done previously.

File: compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java [hunk 9]
```diff
@@ -59,21 +59,17 @@ private static ImmutableMap<String, KtToken> tokenSetToMap(TokenSet tokens) {
         return builder.build();
     }
 
-    private static final TokenSet TYPE_ARGUMENT_LIST_STOPPERS = TokenSet.create(
```
[the remark this answers is not in the document]
Comment (kralliv) [reply compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java:62]: Of course, but I first have to work out how to mimic the behaviour of the old parser :D

File: compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (dzharkov) [reply compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java]: Thanks! It looks just fine to me

File: compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java [hunk 10]
```diff
@@ -484,37 +481,134 @@ else if (atSet(Precedence.POSTFIX.getOperations())) {
...
+        // by some indication of a call suffix or that it's faulty type argument list that
+        // is definitely not a correct right-hand side of an expression and is followed by
+        // a strong indication of a call suffix.
+
+        typeArgumentList.done(TYPE_ARGUMENT_LIST);
 
         return true;
     }
 
+    private boolean isAtTypeArgumentListIndicator() {
```
Comment (dzharkov) [inline compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java:556]: I would rename it to something like `isAtTypeArgumentListSuccessorIndicator` because otherwise one might think from its name that it's about the beginning of argument list.

The same may apply to other similar named-functions below and to the token set fields, too.

File: compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinParsing.java [hunk 11]
```diff
@@ -885,12 +895,10 @@ assert _at(IDENTIFIER) ||
 
         PsiBuilder.Marker reference = mark();
         PsiBuilder.Marker typeReference = mark();
-        parseUserType(/* allowSimpleIntersectionTypes */ false);
+        parseUserType(/* allowSimpleIntersectionTypes */ false,  /* strictTypeArguments */ true);
         typeReference.done(TYPE_REFERENCE);
         reference.done(CONSTRUCTOR_CALLEE);
 
-        parseTypeArgumentList();
```
Comment (dzharkov) [inline compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinParsing.java:892]: While it seemed incorrect, I think it's better to be done in a separate commit with a linked youtrack issue that should be structured in accordance with guidelines: https://kotlinfoundation.org/submitting-incompatible-changes/

But the simplest thing we could do here is just postponing removing that line yet :)

**Review by dzharkov — CHANGES_REQUESTED** [review]

File: compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinParsing.java (hunk 11 above)
Comment (kralliv) [reply compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinParsing.java:892]: I removed it from the PR and I'll report it accordingly.

File: compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java [hunk 12]
```diff
@@ -484,37 +481,143 @@ else if (atSet(Precedence.POSTFIX.getOperations())) {
...
 
+    private boolean isAtTypeArgumentListPostfixIndicator() {
+        return isAtStrongTypeArgumentListPostfixIndicator() || isAtWeakTypeArgumentListPostfixIndicator();
+    }
+
+    private boolean isAtStrongTypeArgumentListPostfixIndicator() {
+        // The start of a value argument list or an annotated lambda indicates a type argument list.
+        // <...>(  <...>{  <...>@identifier{  <...>identifier@{  <...>[
+        // Postfix expression are usually not applied to non-null boolean comparisons.
+        // <...>::  <...>?  <...>.  <...>?.  <...>!!  <...>++  <...>--
```
Comment (dzharkov) [inline compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java:573]: Not sure, about `::` though
Something like `a < b, c > ::d` looks like a reasonable case

File: compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java [hunk 13]
```diff
@@ -484,37 +481,143 @@ else if (atSet(Precedence.POSTFIX.getOperations())) {
...
+        // Other limiting tokens are also a likely indicator of an indented function call.
+        // <...>)  <...>]  <...>}  <...>,  <...>:  <...>=
+        // Operators are unlikely to be applied to non-null booleans.
+        // <...>==  <...>!=  <...>===  <...>!==  <...> as  <...> is  <...> in
+        // Hard keywords hint towards some formatting issue where line breaks are missing
+        // <...>fun  <...>var  <...>var  <...>object
+        return myBuilder.newlineBeforeCurrentToken() || atSet(WEAK_TYPE_ARGUMENT_LIST_POSTFIX_INDICATORS);
+    }
+
+    boolean isAtConditionalExpression(){
```
Comment (dzharkov) [inline compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java:589]: Minor. Missing space before `{`

**Review by dzharkov — CHANGES_REQUESTED** [review]
Sorry, for being late, but those tests are still failing
image

**Comment by kralliv** [comment]
The whole test situation was a bit unfortunate. There were a few tests that admittedly slipped past me, but there was also a change in the FIR compiler, that led to the rightful appearance of an `UNRESOLVED_REFERENCE` error due to a bad interaction with the freshly added `KtNodes.TYPE_ARGUMENT_LIST_LIKE_EXPRESSION` error node. The change was introduced after my push and before you first reported the test failures. Then, in the time in between my next push and your second attempt, that said change was rolled back or broken again and the rightful error was no longer being reported.

The good thing about this unfortunate series of events is, that we got hold of the bad interaction with the freshly introduced error node. The problem was as follows:
The `TYPE_ARGUMENT_LIST_LIKE_EXPRESSION` error node sits in between the left-hand-side of a binary expression and the operator reference.
```
EXPRESSION  TYPE_ARGUMENT_LIST_LIKE_EXPRESSION  OPERATOR_REFERENCE  EXPRESSION
```
The old frontend has no problem with it being there, because it only looks for `KtExpression` elements and `OPERATION_REFERENCE` nodes. However, the FIR is much more generous when it comes to accepting nodes. The `ExpressionConverter.convertBinaryExpression` simply takes the rightmost node left to the operator reference as the left-hand-side, if it is an "expression" by FIR standards. The `TYPE_ARGUMENT_LIST_LIKE_EXPRESSION` error node qualified as an expression, because it was of type `KtNode`.
Now, I am not qualified to judge over what counts as an expression in FIR, so I made no change to the conversion. I instead changed the type of `TYPE_ARGUMENT_LIST_LIKE_EXPRESSION` from `KtNode` to `IElementType`. I don't know, if there is a better type to be used here instead.
In the future that conversion probably has to improve. For instance, right now, it parses the right-hand-side as the rightmost (and NOT the leftmost) expression node to the right of the operator reference.

File: compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java (hunk 12 above)
Comment (kralliv) [reply compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java:573]: Without it, code like `Array<String>::class` or `foo<String, String>::bar::com` would not parse correctly. We cannot know how many arguments the list actually takes, given that it would require type resolution, which is why I would say, that type argument lists have to take precedence.

**Review by dzharkov — CHANGES_REQUESTED** [review]
Sorry for the delay, but it seems that we're close to the wrapping final things up.
Also, please, rebase the branch again.

File: compiler/testData/psi/CallsInWhen_ERR.kt [hunk 14]
```diff
@@ -0,0 +1,7 @@
+fun foo() {
+  when (a) {
+    a.foo<T> -> a
```
[the remark this answers is not in the document]
Comment (kralliv) [reply compiler/testData/psi/CallsInWhen_ERR.kt:3]: `when` expression allow for writing code like the following:
```kotlin
when {
    a < b, b > c -> TODO()
}
```
Those aren't even unlikely cases. Making a solid suggestion would require a lot more fine grained introspection of all the different cases. IntelliJ autocompletes function calls with `()` at the end, which in turn means that the compiler will consider it as a type argument list. This should strike somewhat of a balance between typing function calls and comparisons.

[the remark this answers is not in the document]
Comment (kralliv) [reply compiler/testData/psi/CallsInWhen_ERR.kt:3]: The problem that I see is, that any incomplete expression would never be successfully parsed. But, what we could do the following:

Currently we do this:
```java
if (kind == KotlinParsing.TypeArgumentListKind.FAULTY_TYPE_ARGUMENT_LIST) {
    if (!isAtStrongTypeArgumentListPostfixIndicator()) {
        typeArgumentList.rollbackTo();

        mark().done(TYPE_ARGUMENT_LIST_LIKE_EXPRESSION);
        return false;
    }

    typeArgumentList.rollbackTo();

    if (isAtConditionalExpression()) {
        mark().done(TYPE_ARGUMENT_LIST_LIKE_EXPRESSION);
        return false;
    }

    typeArgumentList = mark();

    myKotlinParsing.tryParseTypeArgumentList();
}
```
We could change it to something like this:
```java
if (kind == KotlinParsing.TypeArgumentListKind.FAULTY_TYPE_ARGUMENT_LIST) {
    int commaCount = getCommaCount(typeArgumentList);

    typeArgumentList.rollbackTo();

    if (isAtConditionalExpression() || !isAtStrongTypeArgumentListPostfixIndicator() && commaCount == 1) {
        mark().done(TYPE_ARGUMENT_LIST_LIKE_EXPRESSION);
        return false;
    }

    typeArgumentList = mark();

    myKotlinParsing.tryParseTypeArgumentList();
}
```
where `getCommaCount(typeArgumentList)` counts the number of commas in the type argument list.

This would change nothing about the correctness, but would lean a lot more towards call suffixes, except when the type argument list contains exactly contains one comma. It wouldn't handle writing expressions like `a < b > false` well, but those are kinda nonsense anyways. And due to line break being a strong indicator, cases like top level `mutableMapOf<String, String>` would still be considered a call suffix despite containing exactly one comma.

[the remark this answers is not in the document]
Comment (dzharkov) [reply compiler/testData/psi/CallsInWhen_ERR.kt:3]: To be honest, while mostly that trick with `getCommaCount` looks reasonable, I'd prefer to avoid another level of complexity here.
So, I'd suggest to add `->` to the list of weak call postfix comma, so we'd have a conservative solution that preserves the behavior for `a<b> ->` case.
Next time, once we get some complaints about that behavior or would be more decisive to add another layer of recovery logic, we may introduce it.

[the remark this answers is not in the document]
Comment (kralliv) [reply compiler/testData/psi/CallsInWhen_ERR.kt:3]: I guess that would make for good compromise. It would hurt the autocompletion of something like the following, but correctness would be still be given. And as soon as there is an error in the type argument list, it would fall back to a comparison again, so it should be ok.
```kotlin
when {
    a < b,
    a > ->TODO()
}
```

[the remark this answers is not in the document]
Comment (dzharkov) [reply compiler/testData/psi/CallsInWhen_ERR.kt]: Sorry, I didn't understand, but may you elaborate what you mean by "that" here?

[the remark this answers is not in the document]
Comment (kralliv) [reply compiler/testData/psi/CallsInWhen_ERR.kt]: My bad, I meant your idea to add the `->` to the weak indicator set. I initially only considered the strong indicator set, when I proposed the `getCommaCount()` approach, but that would make it lean too much towards call suffixes. The weak indicator set, however, should make for a good compromise. I added the `ARROW` to the weak indicator set and adjusted the tests.

**Comment by dzharkov** [comment]
Now, it's almost done!
But unfortunately the test `org.jetbrains.kotlin.test.runners.DiagnosticTestGenerated$Tests$TypeArguments.testForbidTypeArgumentListLikeExpressions` is failing

**Comment by kralliv** [comment]
Hm, doesn't the Gradle run configuration "Analysis All Tests" execute all tests? It seems to not do that, at least not all of the time.
Anyway, I forgot to adjust the `TypeArgumentListLikeExpressionsChecker` accordingly after I changed the type of the `KtNodeTypes.TYPE_ARGUMENT_LIST_LIKE_EXPRESSION`.

**Comment by dzharkov** [comment]
I merged your changes manually and then unfortunately had to revert them because IDE plugin tests start failing :(
The tests are the following
image

It seems that after your changes, the code like `HashMap<$$some-special-character-for-completion`  starts being parsed as a condition instead of a type argument list that leads to redundant suggestions of functions instead of types.
Sorry, I don't have resources now for deeper investigation, I could only suggest to reproduce it via setuping [Cooperative project](https://github.com/JetBrains/intellij-community/blob/master/plugins/kotlin/docs/cooperative-development/environment-setup.md) locally.

**Comment by kralliv** [comment]
No problem. I have already checked out the intellij repository and I'll take a look at the issue.

**Comment by kralliv** [comment]
After understanding what Intellij does to perform the completion of incomplete statements, I adjusted the strong indicator list to contain `TokenType.BAD_CHARACTER`. With this change the mentioned tests now run.
For context:
Intellij completes code like `HashMap<String,` with `HashMap<String, IntellijIdeaRulezzz>$` in an attempt to get it to compile. `$` is considered a bad character everywhere except in strings and in front of identifiers (legacy syntax). Previously the bad character didn't matter, but now it must be considered as an indicator of a type argument list in order to conform to the existing assumptions.

**Comment by dzharkov** [comment]
Merged your changes manually, it seems like everything went fine.
Thank you for your work and patience when fixing these complicated issues!
